hypervisor/arch/x86/virq.c

```diff
@@ -469,11 +448,32 @@ int acrn_handle_pending_request(struct vcpu *vcpu)
 		goto INTR_WIN;
 
 INTR_WIN:
-	/* check if we have new interrupt pending for next VMExit */
-	intr_pending = vcpu_pending_request(vcpu);
+	/* The pending ExtInt and vLAPIC vectors may hasn't inject due
+	 * to blocking by RFLAGS.IF = 0 or STI or MOV SS.
+	 *
+	 * And external interrupt(from vpic) can't be delivered by "virtual-
+	 * interrupt delivery", "virtual-interrupt delivery" only deliver
+	 * interrupt from vlapic.
+	 *
+	 * So need to enable "interrupt-window exiting", when there is
+	 * an ExtInt, or "virtual-interrupt delivery" is disabled and there
+	 * is lapic interrupt which failed inject in this time.
+	 */
+	if ((arch_vcpu->irq_window_enabled == 1U) ||
+		(is_guest_irq_enabled(vcpu) == true)) {
+		return ret;
+	}
+
+	if ((bitmap_test(ffs64(ACRN_REQUEST_EXTINT),
+						pending_req_bits) == 0U) &&
+		((is_apicv_intr_delivery_supported() == true) ||
+		 (vlapic_pending_intr(vcpu->arch_vcpu.vlapic,
+							  &vector) == 0U))) {
+		return ret;
+	}
 
 	/* Enable interrupt window exiting if pending */
-	if (intr_pending && arch_vcpu->irq_window_enabled == 0U) {
+	if (arch_vcpu->irq_window_enabled == 0U) {
```

Comment:
if (arch_vcpu->irq_window_enabled == 0U) {
this can be removed, it is checked here
 if ((arch_vcpu->irq_window_enabled == 1U) ||
		(is_guest_irq_enabled(vcpu) == true)) {
		return ret;
	}

Reply:
Correct. I will remove it